File: compile.py
#!/usr/bin/python

import os
import sys
import pathlib
import platform
import shutil

PRO = "cgraph"
DIR = "."
SRC = os.path.join(DIR, "src")
INC = os.path.join(DIR, "include")
TST = os.path.join(DIR, "test")
LIB = os.path.join(DIR, "lib")

CC = "cc"
CFLAGS = "-ansi -pedantic -pedantic-errors -Wall -fPIC -g"
CSFLAGS = "-shared"

# package shared library
AR = "ar"
ARFLAGS = "-rcs"

plat = platform.system()
if plat == "Windows":
  # target files
  LIBSHARED = os.path.join(LIB, "lib%s.dll" % PRO)
  LIBSTATIC = os.path.join(LIB, "lib%s.a" % PRO)
  # test files
  TSTFILE = os.path.join(TST, "%s.c" % PRO)
  TSTTARGET = os.path.join(TST, "%s.exe" % PRO)
else:
  # target files
  LIBSHARED = os.path.join(LIB, "lib%s.so" % PRO)
  LIBSTATIC = os.path.join(LIB, "lib%s.a" % PRO)
  # test files
  TSTFILE = os.path.join(TST, "%s.c" % PRO)
  TSTTARGET = os.path.join(TST, PRO)

CFILES = []
for file in os.listdir(SRC):
  if os.path.splitext(file)[1] == ".c":
    CFILES.append(file)

args = sys.argv
if len(args) == 1 :
  if not pathlib.Path(LIB).exists():
    os.mkdir(LIB)
  for file in CFILES:
    obj = os.path.splitext(file)[0] + ".o"
    file_path = os.path.join(SRC, file)
    obj_path = os.path.join(SRC, obj)
    print("compile %s to %s" % (file_path, obj_path))
    os.system("%s %s -I%s -c %s -o %s" % (CC, CFLAGS, INC, file_path, obj_path))
  print("compile %s" % LIBSHARED)
  os.system("%s %s -o %s %s/*.o" % (CC, CSFLAGS, LIBSHARED, SRC))
  print("compile %s" % LIBSTATIC)
  os.system("%s %s %s %s/*.o" % (AR, ARFLAGS, LIBSTATIC, SRC))
elif args[1] == "test":
  print("compile %s to %s" % (os.path.join(TST, TSTFILE), os.path.join(TST, TSTTARGET)))
  os.system("%s %s -I%s -o %s %s -L%s -static -l%s -lm" % (CC, CFLAGS, INC, TSTTARGET, TSTFILE, LIB, PRO))
  print("test %s with %s" % (TSTTARGET, os.path.join(TST, "elements.csv")))
  os.system("%s %s" % (TSTTARGET, os.path.join(TST, "elements.csv")))
elif args[1] == "clean":
  for file in CFILES:
    obj = os.path.splitext(file)[0] + ".o"
    file_path = os.path.join(SRC, file)
    obj_path = os.path.join(SRC, obj)
    print("clean %s" % obj_path)
    if pathlib.Path(obj_path).exists():
      os.remove(obj_path)
  print("clean %s" % LIBSTATIC)
  if pathlib.Path(LIBSTATIC).exists():
    os.remove(LIBSTATIC)
  print("clean %s" % LIBSHARED)
  if pathlib.Path(LIBSHARED).exists():
    os.remove(LIBSHARED)
  print("clean %s" % TSTTARGET)
  if pathlib.Path(TSTTARGET).exists():
    os.remove(TSTTARGET)
elif args[1] == "distclean":
  for file in CFILES:
    obj = os.path.splitext(file)[0] + ".o"
    file_path = os.path.join(SRC, file)
    obj_path = os.path.join(SRC, obj)
    print("clean %s" % obj_path)
    if pathlib.Path(obj_path).exists():
      os.remove(obj_path)
  print("clean %s" % LIBSTATIC)
  if pathlib.Path(LIBSTATIC).exists():
    os.remove(LIBSTATIC)
  print("clean %s" % LIBSHARED)
  if pathlib.Path(LIBSHARED).exists():
    os.remove(LIBSHARED)
  print("clean %s" % LIB)
  if pathlib.Path(LIB).exists():
    shutil.rmtree(LIB)
  print("clean %s" % TSTTARGET)
  if pathlib.Path(TSTTARGET).exists():
    os.remove(TSTTARGET)
elif args[1] == "help":
  print("%s <target>" % args[0])
  print("<target>: ")
  print("                    compile cgraph")
  print("          test      test cgraph")
  print("          clean     clean all the generated files")
  print("          distclean clean all the generated files and directories")
  print("          help      commands to this program")
else:
  print("%s is an unsupported command" % args[0])
  print("use \"%s help\" to know all supported commands" % args[0])

File: compile.r
#!/usr/bin/Rscript

PRO <- "cgraph"
DIR <- "."
SRC <- file.path(DIR, "src", fsep=.Platform$file.sep)
INC <- file.path(DIR, "include", fsep=.Platform$file.sep)
TST <- file.path(DIR, "test", fsep=.Platform$file.sep)
LIB <- file.path(DIR, "lib", fsep=.Platform$file.sep)

CC <- "cc"
CFLAGS <- "-pedantic -Wall -fpic -std=c89"
CSFLAGS <- "-shared"

MODE <- "debug"
if(MODE == "debug") {
  CFLAGS <- paste(CFLAGS, "-g -DDEBUG", sep=" ")
} else if(MODE == "release") {
  CFLAGS <- paste(CFLAGS, "-static -O2", sep=" ")
}

# package shared library
AR <- "ar"
ARFLAGS <- "-rcs"

if(.Platform$OS.type == "windows") {
  # target files
  LIBSHARED <- file.path(LIB, paste("lib", PRO, ".dll", sep=""), fsep=.Platform$file.sep)
  LIBSTATIC <- file.path(LIB, paste("lib", PRO, ".a", sep=""), fsep=.Platform$file.sep)
  # test files
  TSTFILE <- file.path(TST, paste(PRO, ".c", sep=""), fsep=.Platform$file.sep)
  TSTTARGET <- file.path(TST, paste(PRO, ".exe", sep=""), fsep=.Platform$file.sep)
} else {
  # target files
  LIBSHARED <- file.path(LIB, paste("lib", PRO, ".so", sep=""), fsep=.Platform$file.sep)
  LIBSTATIC <- file.path(LIB, paste("lib", PRO, ".a", sep=""), fsep=.Platform$file.sep)
  # test files
  TSTFILE <- file.path(TST, paste(PRO, ".c", sep=""), fsep=.Platform$file.sep)
  TSTTARGET <- file.path(TST, PRO, fsep=.Platform$file.sep)
}

CFILES <- list.files(SRC, pattern=".c$")

args <- commandArgs(trailingOnly = TRUE)
script_name <- unlist(strsplit(commandArgs()[4], split="="))[2]
if (length(args) == 0) {
  dir.create(LIB)
  for(file in CFILES) {
    obj <- gsub(".c$", ".o", file)
    print(sprintf("compile %s to %s", file.path(SRC, file, fsep=.Platform$file.sep), file.path(SRC, obj, fsep=.Platform$file.sep)))
    system(sprintf("%s %s -I%s -c %s -o %s", CC, CFLAGS, INC, file.path(SRC, file, fsep=.Platform$file.sep), file.path(SRC, obj, fsep=.Platform$file.sep)))
  }
  print(sprintf("compile %s", LIBSHARED))
  system(sprintf("%s %s -o %s %s/*.o", CC, CSFLAGS, LIBSHARED, SRC))
  print(sprintf("compile %s", LIBSTATIC))
  system(sprintf("%s %s %s %s/*.o", AR, ARFLAGS, LIBSTATIC, SRC))
} else if (args[1] == "test") {
  print(sprintf("compile %s to %s", TSTFILE, TSTTARGET))
  system(sprintf("%s %s -I%s -o %s %s -L%s -static -l%s -lm", CC, CFLAGS, INC, TSTTARGET, TSTFILE, LIB, PRO))
  print(sprintf("test %s with %s", TSTTARGET, file.path(TST, "elements.csv", fsep=.Platform$file.sep)))
  system(sprintf("%s %s", TSTTARGET, file.path(TST, "elements.csv", fsep=.Platform$file.sep)))
} else if (args[1] == "clean") {
  for(file in CFILES) {
    obj <- gsub(".c$", ".o", file)
    print(sprintf("clean %s", obj))
    unlink(file.path(SRC, obj, fsep=.Platform$file.sep), force=TRUE)
  }
  print(sprintf("clean %s", LIBSTATIC))
  unlink(LIBSTATIC, force=TRUE)
  print(sprintf("clean %s", LIBSHARED))
  unlink(LIBSHARED, force=TRUE)
  print(sprintf("clean %s", TSTTARGET))
  unlink(TSTTARGET, force=TRUE)
} else if (args[1] == "distclean") {
  for(file in CFILES) {
    obj <- gsub(".c$", ".o", file)
    print(sprintf("clean %s", obj))
    unlink(file.path(SRC, obj, fsep=.Platform$file.sep), force=TRUE)
  }
  print(sprintf("clean %s", LIBSTATIC))
  unlink(LIBSTATIC, force=TRUE)
  print(sprintf("clean %s", LIBSHARED))
  unlink(LIBSHARED, force=TRUE)
  print(sprintf("clean %s", LIB))
  unlink(LIB, force=TRUE)
  print(sprintf("clean %s", TSTTARGET))
  unlink(TSTTARGET, force=TRUE)
} else if (args[1] == "help") {
  print(sprintf("%s <target>", script_name))
  print("<target>: ")
  print("                    compile cgraph")
  print("          test      test cgraph")
  print("          clean     clean all the generated files")
  print("          distclean clean all the generated files and directories")
  print("          help      commands to this program")
} else {
  print(sprintf("%s is an unsupported command", args[1]))
  print(sprintf("use \"%s help\" to know all supported commands", script_name))
}